### PR TITLE
Fix dropdown escape propagation

### LINKED ISSUE #33465: v5: Dropdown trigger button "swallows" Esc keypresses when focused
Split off from https://github.com/twbs/bootstrap/issues/23507#issuecomment-805849754 (which is an issue about v4).

If a modal dialog contains a dropdown, `Esc` is swallowed/intercepted by the dropdown trigger button even when the dropdown itself is closed.

Instead, the keypress should not be handled when the dropdown is closed, letting it bubble up to the modal.

See the demo https://codepen.io/Jarek3000/pen/LYxVJJg

* launch the modal
* `Tab` to the dropdown button so it receives focus
* press `Esc` - nothing happens, as the keypress is intercepted by the modal code

### PR DESCRIPTION
**Pull request by alpadev**
Closes #33465

### CODE REVIEW

**Comment by GeoSot** [comment]
LGTM

File: js/src/dropdown.js [hunk 1]
```diff
@@ -463,26 +488,29 @@ class Dropdown extends BaseComponent {
...
     event.preventDefault()
     event.stopPropagation()
 
     if (this.disabled || this.classList.contains(CLASS_NAME_DISABLED)) {
       return
     }
 
-    const parent = Dropdown.getParentFromElement(this)
-    const isActive = this.classList.contains(CLASS_NAME_SHOW)
+    const getButton = () => this.matches(SELECTOR_DATA_TOGGLE) ? this : SelectorEngine.prev(this, SELECTOR_DATA_TOGGLE)[0]
```
Comment (rohit2sharma95) [inline js/src/dropdown.js:504]: This may be better :slightly_smiling_face: 
`getButton` => `getToggleButton`

**Review by rohit2sharma95 — COMMENTED** [review]
https://github.com/twbs/bootstrap/blob/5cc53a0ef0718b43e376462f156d39bf5542fbf9/js/src/dropdown.js#L489-L492

This should not be changed as below? :thinking: 
```js
if (isActive && event.key === SPACE_KEY) {
  Dropdown.clearMenus()
  return
}
```

**Comment by alpadev** [comment]
> https://github.com/twbs/bootstrap/blob/5cc53a0ef0718b43e376462f156d39bf5542fbf9/js/src/dropdown.js#L489-L492
> 
> This should not be changed as below? 🤔
> 
> ```js
> if (isActive && event.key === SPACE_KEY) {
>   Dropdown.clearMenus()
>   return
> }
> ```

Not sure if this is even needed.

This checks for keys that are dropdown commands.

https://github.com/twbs/bootstrap/blob/5cc53a0ef0718b43e376462f156d39bf5542fbf9/js/src/dropdown.js#L458-L464

But the regex is missing the space key so I'm not sure if the code would even reach this block.

https://github.com/twbs/bootstrap/blob/5cc53a0ef0718b43e376462f156d39bf5542fbf9/js/src/dropdown.js#L44

**Comment by rohit2sharma95** [comment]
This change was made in #29885. Anything related to input groups?

/CC @MartijnCuppens

**Comment by rohit2sharma95** [comment]
And can not the `selectMenuItem` method be a private method instead of a static one (as it is not being used directly :thinking:)? It can be used through the context of the dropdown then IMO. Because the `dataApiKeydownHandler` method is bound with the dropdown elements only (menu and the toggle button).

**Comment by alpadev** [comment]
> And can not the `selectMenuItem` method be a private method instead of a static one (as it is not being used directly 🤔)? It can be used through the context of the dropdown then IMO. Because the `dataApiKeydownHandler` method is bound with the dropdown elements only (menu and the toggle button).

I didn't mean to refactor too much in the scope of this issue but rather to just get rid of that Eslint warning. I agree that using a static method isn't ideal but the item slection functionality wasn't bound to the instance before and I made it the way `getParentFromElement` was implemented.

If you want me to change it I should be able to do this tho.

**Comment by GeoSot** [comment]
Till now is just a refactoring that I can follow, a minor fix and a valid test that succeeds . So **till here** I find it **OK**.

I agree to refactor `selectMenuItem` to private method, and would also love to see `dataApiKeydownHandler` gordian knot untied, but on another PR, to avoid mistakes

So if we can all agree on these, I would prefer to keep it as it is and if @alpadev has the courage to keep up with a following PR

**Review by GeoSot — APPROVED** [review]

**Pull request merged**